File: paimon-core/src/main/java/org/apache/paimon/operation/metrics/ScanStats.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.paimon.operation.metrics;

import org.apache.paimon.annotation.VisibleForTesting;

/** Statistics for a scan operation. */
public class ScanStats {
    // the unit is milliseconds
    private final long duration;
    private final long scannedManifests;
    private final long skippedByPartitionAndStats;
    private final long skippedByBucketAndLevelFilter;

    private final long skippedByWholeBucketFiles;
    private final long skippedTableFiles;
    private final long resultedTableFiles;

    public ScanStats(
            long duration,
            long scannedManifests,
            long skippedByPartitionAndStats,
            long skippedByBucketAndLevelFilter,
            long skippedByWholeBucketFiles,
            long resultedTableFiles) {
        this.duration = duration;
        this.scannedManifests = scannedManifests;
        this.skippedByPartitionAndStats = skippedByPartitionAndStats;
        this.skippedByBucketAndLevelFilter = skippedByBucketAndLevelFilter;
        this.skippedByWholeBucketFiles = skippedByWholeBucketFiles;
        this.skippedTableFiles =
                skippedByPartitionAndStats
                        + skippedByBucketAndLevelFilter
                        + skippedByWholeBucketFiles;
        this.resultedTableFiles = resultedTableFiles;
    }

    @VisibleForTesting
    protected long getScannedManifests() {
        return scannedManifests;
    }

    @VisibleForTesting
    protected long getSkippedTableFiles() {
        return skippedTableFiles;
    }

    @VisibleForTesting
    protected long getResultedTableFiles() {
        return resultedTableFiles;
    }

    @VisibleForTesting
    protected long getSkippedByPartitionAndStats() {
        return skippedByPartitionAndStats;
    }

    @VisibleForTesting
    protected long getSkippedByBucketAndLevelFilter() {
        return skippedByBucketAndLevelFilter;
    }

    @VisibleForTesting
    protected long getSkippedByWholeBucketFiles() {
        return skippedByWholeBucketFiles;
    }

    @VisibleForTesting
    protected long getDuration() {
        return duration;
    }
}

File: paimon-python/pypaimon/read/scan_stats.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

from dataclasses import dataclass, field
from typing import Set, Tuple


@dataclass
class ScanStats:
    """Counters accumulated by :class:`FileScanner` when it is asked to
    track stats for ``ReadBuilder.explain()``.

    The scanner mutates these counters in place; consumers should treat
    instances as immutable once the scan returns. Default factory values
    keep the dataclass usable both as a blank "no tracking" sentinel and
    as a live accumulator.
    """

    manifest_files_total: int = 0
    manifest_files_after_partition: int = 0

    # ``entries_potential_total`` is the row count we would have processed if no
    # filtering occurred — derived from manifest-file metadata before the
    # manifest-level partition skip. ``entries_total`` is what actually reached
    # ``_filter_manifest_entry``; the gap between the two captures
    # manifest-level pruning.
    entries_potential_total: int = 0
    entries_total: int = 0
    entries_after_partition: int = 0
    entries_after_bucket: int = 0
    entries_after_stats: int = 0

    partition_keys_before: Set[Tuple] = field(default_factory=set)
    partition_keys_after: Set[Tuple] = field(default_factory=set)

    buckets_seen: Set[Tuple[Tuple, int]] = field(default_factory=set)
    buckets_after_pruning: Set[Tuple[Tuple, int]] = field(default_factory=set)
